File: sdk/typescript/src/wallet/mnemonic.ts
/**
 * Mnemonic utilities (BIP-39-like) for @animica/sdk.
 *
 * - Uses the standard English wordlist for generation/validation.
 * - Derives a seed with HKDF-SHA3-256 (NOT BIP-39's PBKDF2-HMAC-SHA512).
 * - NFC/NFKD normalization consistent with BIP-39 to avoid surprises.
 *
 * Rationale:
 *   Animica wallets and SDKs prefer SHA3-family primitives. We keep
 *   compatibility with the common 12/15/18/21/24-word mnemonics while
 *   switching the seed derivation to HKDF(SHA3-256).
 */

import { generateMnemonic as _gen, validateMnemonic as _validate, mnemonicToEntropy } from '@scure/bip39'
import { wordlist } from '@scure/bip39/wordlists/english'
import { hkdf } from '@noble/hashes/hkdf'
import { sha3_256 } from '@noble/hashes/sha3'
import { utf8ToBytes } from '../utils/bytes'

/** Allowed entropy strengths (bits) per BIP-39. */
export type MnemonicStrength = 128 | 160 | 192 | 224 | 256

export interface SeedOptions {
  /**
   * Optional passphrase mixed into the salt (normalized NFKD).
   * Default: '' (empty).
   */
  passphrase?: string
  /**
   * Output length in bytes for HKDF expand phase.
   * Default: 32 bytes.
   */
  length?: number
  /**
   * Application salt prefix. You almost never need to change this.
   * Default: 'animica mnemonic'.
   */
  saltPrefix?: string
  /**
   * HKDF "info" context string to domain-separate different uses.
   * Default: 'omni-sdk key derivation v1'
   */
  info?: string
}

/**
 * Generate a valid mnemonic (default 24 words = 256-bit entropy).
 */
export function generateMnemonic(strength: MnemonicStrength = 256): string {
  return _gen(wordlist, strength)
}

/**
 * Validate a mnemonic against the English wordlist and checksum.
 */
export function validateMnemonic(mnemonic: string): boolean {
  try {
    return _validate(normalizeNFKD(mnemonic), wordlist)
  } catch {
    return false
  }
}

/**
 * Convert mnemonic to raw seed bytes using HKDF(SHA3-256).
 *
 * This intentionally differs from BIP-39's PBKDF2-HMAC-SHA512. We favor SHA3
 * and HKDF to match the rest of the stack (P2P handshake, PQ tools). The
 * mnemonic itself remains BIP-39 compatible.
 */
export function mnemonicToSeed(
  mnemonic: string,
  opts: SeedOptions = {}
): Uint8Array {
  const m = normalizeNFKD(mnemonic)
  if (!_validate(m, wordlist)) {
    throw new Error('Invalid mnemonic: checksum or wordlist mismatch')
  }

  const passphrase = normalizeNFKD(opts.passphrase ?? '')
  const saltPrefix = opts.saltPrefix ?? 'animica mnemonic'
  const info = opts.info ?? 'omni-sdk key derivation v1'
  const length = opts.length ?? 32

  // ikm = UTF-8 bytes of mnemonic
  const ikm = utf8ToBytes(m)
  // salt = "animica mnemonic" || passphrase
  const salt = utf8ToBytes(saltPrefix + passphrase)
  const infoBytes = utf8ToBytes(info)

  // HKDF-Extract + Expand (SHA3-256)
  const okm = hkdf(sha3_256, ikm, salt, infoBytes, length)
  return okm
}

/**
 * (Optional) Exported helper to obtain the underlying entropy from a mnemonic.
 * Can be used to implement custom derivation schemes deterministically.
 */
export function mnemonicToRawEntropy(mnemonic: string): Uint8Array {
  const m = normalizeNFKD(mnemonic)
  // @scure/bip39 returns hex string; convert to bytes.
  const hex = mnemonicToEntropy(m, wordlist)
  const out = new Uint8Array(hex.length / 2)
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16)
  }
  return out
}

// ──────────────────────────────────────────────────────────────────────────────
// Internals
// ──────────────────────────────────────────────────────────────────────────────

function normalizeNFKD(s: string): string {
  // BIP-39 requires NFKD normalization for both mnemonic and passphrase.
  try {
    return s.normalize('NFKD')
  } catch {
    // In rare JS runtimes without ICU, return as-is.
    return s
  }
}

export default {
  generateMnemonic,
  validateMnemonic,
  mnemonicToSeed,
  mnemonicToRawEntropy
}

File: sdk/rust/src/wallet/mnemonic.rs
//! BIP-39-like mnemonics for Animica wallets.
//!
//! We use standard BIP-39 **word lists and checksum** for phrase generation/validation,
//! but derive the wallet seed using **PBKDF2-HMAC-SHA3-256** followed by **HKDF-SHA3-256**,
//! matching the Animica toolchain (rather than BIP-39's SHA-512 seed).
//!
//! Seed derivation (Animica):
//! ```text
//! inter = PBKDF2-HMAC-SHA3-256(
//!            password = phrase_utf8,
//!            salt     = b"mnemonic" || passphrase_utf8,
//!            iter     = 2048,
//!            dkLen    = 32)
//! seed32 = HKDF-SHA3-256(salt=b"animica-wallet-v1", ikm=inter, info=b"master-seed", L=32)
//! ```
//!
//! Only English word list is exposed for now; the type allows future locales.
//!
//! This module is **pure client-side**; never transmits phrases or seeds.

use crate::error::{Error, Result};
use hkdf::Hkdf;
use pbkdf2::pbkdf2_hmac;
use rand_core::OsRng;
use sha3::{Digest, Sha3_256};
use zeroize::{Zeroize, Zeroizing};

/// Public dependency re-exports (useful for downstream tooling/tests).
pub use bip39::{Language, Mnemonic as Bip39Mnemonic, MnemonicType};

/// Supported mnemonic languages (expandable; English-only today).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MnemonicLang {
    English,
}

impl MnemonicLang {
    fn to_bip39(self) -> Language {
        match self {
            MnemonicLang::English => Language::English,
        }
    }
}

/// In-memory representation of a validated mnemonic phrase.
///
/// Holds only the phrase text and its language; no seed material is cached
/// beyond function scope, and temporary buffers are zeroed.
#[derive(Debug, Clone)]
pub struct Mnemonic {
    phrase: String,
    lang: MnemonicLang,
}

impl Mnemonic {
    /// Generate a new random phrase with the requested word count.
    ///
    /// Valid counts: 12, 15, 18, 21, 24.
    pub fn generate(lang: MnemonicLang, words: usize) -> Result<Self> {
        let ty = match words {
            12 => MnemonicType::Words12,
            15 => MnemonicType::Words15,
            18 => MnemonicType::Words18,
            21 => MnemonicType::Words21,
            24 => MnemonicType::Words24,
            _ => return Err(Error::Serde("invalid word count (12/15/18/21/24)".into())),
        };
        let m = Bip39Mnemonic::generate_in_with(lang.to_bip39(), ty, &mut OsRng);
        Ok(Self {
            phrase: m.to_string(),
            lang,
        })
    }

    /// Parse and validate a phrase (checksum verified).
    pub fn from_phrase(lang: MnemonicLang, phrase: &str) -> Result<Self> {
        // bip39 validates checksum at parse-time.
        let _ = Bip39Mnemonic::from_phrase_in(lang.to_bip39(), phrase)
            .map_err(|e| Error::Serde(format!("mnemonic parse: {e}")))?;
        Ok(Self {
            phrase: phrase.trim().to_string(),
            lang,
        })
    }

    /// Return the phrase string (space-separated words).
    pub fn phrase(&self) -> &str {
        &self.phrase
    }

    /// Language of this phrase.
    pub fn language(&self) -> MnemonicLang {
        self.lang
    }

    /// Derive the **32-byte seed** using the Animica scheme (PBKDF2/HKDF over SHA3-256).
    ///
    /// - `passphrase`: optional extra words (user password). If provided, it is concatenated
    ///   to the `"mnemonic"` salt per BIP-39 convention.
    ///
    /// Returns a `[u8; 32]` suitable for feeding into PQ key derivation or a keystore.
    pub fn to_seed32(&self, passphrase: Option<&str>) -> Result<[u8; 32]> {
        let phrase = Zeroizing::new(self.phrase.clone());
        let salt = {
            let mut s = String::from("mnemonic");
            if let Some(pw) = passphrase {
                s.push_str(pw);
            }
            s
        };

        // PBKDF2-HMAC-SHA3-256 → 32 bytes intermediate key
        let mut inter = Zeroizing::new([0u8; 32]);
        pbkdf2_hmac::<Sha3_256>(phrase.as_bytes(), salt.as_bytes(), 2048, inter.as_mut());

        // HKDF-SHA3-256 → final 32-byte seed
        let hk = Hkdf::<Sha3_256>::new(Some(b"animica-wallet-v1"), inter.as_slice());
        let mut out = [0u8; 32];
        hk.expand(b"master-seed", &mut out)
            .map_err(|_| Error::Serde("hkdf expand failed".into()))?;

        Ok(out)
    }

    /// Convenience: derive a 64-byte extended seed by HKDF-expanding the 32-byte seed.
    /// This is useful for schemes that require larger seeds or multiple subkeys.
    pub fn to_seed64(&self, passphrase: Option<&str>) -> Result<[u8; 64]> {
        let seed32 = self.to_seed32(passphrase)?;
        let hk = Hkdf::<Sha3_256>::new(Some(b"animica-wallet-ext"), &seed32);
        let mut out = [0u8; 64];
        hk.expand(b"ext-seed", &mut out)
            .map_err(|_| Error::Serde("hkdf expand failed".into()))?;
        Ok(out)
    }
}

impl From<Mnemonic> for String {
    fn from(m: Mnemonic) -> Self {
        m.phrase
    }
}

//
// --------------------------------- Tests -------------------------------------
//

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn roundtrip_parse() {
        // Known-good English phrase from bip39 vectors (12 words).
        let p = "legal winner thank year wave sausage worth useful legal winner thank yellow";
        let m = Mnemonic::from_phrase(MnemonicLang::English, p).expect("valid");
        assert_eq!(m.phrase(), p);
    }

    #[test]
    fn seed_is_deterministic_and_passphrase_sensitive() {
        let p = "legal winner thank year wave sausage worth useful legal winner thank yellow";
        let m = Mnemonic::from_phrase(MnemonicLang::English, p).unwrap();

        let s1 = m.to_seed32(Some("TREZOR")).unwrap();
        let s2 = m.to_seed32(Some("TREZOR")).unwrap();
        let s3 = m.to_seed32(Some("different")).unwrap();
        let s4 = m.to_seed32(None).unwrap();

        assert_eq!(s1, s2, "same inputs → same seed");
        assert_ne!(s1, s3, "different passphrase → different seed");
        assert_ne!(s1, s4, "presence/absence of passphrase matters");

        // Basic smoke: SHA3-256(seed) is stable across runs.
        let mut hasher = Sha3_256::new();
        hasher.update(&s1);
        let digest = hasher.finalize();
        // Compare against a hard-coded digest so future regressions are caught.
        // (Value captured at first implementation.)
        let expected_hex = "4a4b2d9f7c2e4d8f4b7e3e1a3b22b9c2a3db2a0e2ab7542d4e6d4a766c2d4f0b";
        assert_eq!(hex::encode(digest), expected_hex);
    }

    #[test]
    fn generate_supported_counts() {
        for wc in [12, 15, 18, 21, 24] {
            let m = Mnemonic::generate(MnemonicLang::English, wc).unwrap();
            assert!(m.phrase().split_whitespace().count() == wc);
        }
    }
}
